### PR TITLE
Add throwExceptionTx and mapExceptionTx functions

### PR DESCRIPTION
**Pull request by jship**
This PR adds `throwExceptionTx` and `mapExceptionTx` functions.  For `mapExceptionTx`, it also adds some typeclass infra (`UnsafeUnliftTx`).

---

~This PR adds:~ Outdated as of 08/06/2020

- ~`unsafeWithRunInIOTxM` function. This is useful at the top of a `TxM` setup to be able to acquire a connection in `IO`.~
- ~`MonadCatch`, `MonadThrow`, and `MonadMask` instances to `TxM`. If any code in `PgQueryM` for example, was using `MonadCatch`'s `catch`, because `PgQueryM` has `TxM` at the base, this ensures that code is still compatible.~

### CODE REVIEW

**Comment by jfischoff** [comment]
Something to consider with `MonadCatch`: without using savepoints you will not rollback the statements that occurred before the exception was thrown. 

I'm not sure there is necessarily a way around this with this monad. Here is way I handled it for a postgresql-simple transaction monad: https://github.com/jfischoff/pg-transact/blob/master/src/Database/PostgreSQL/Transact.hs#L53

This is similar to how Postgres implements exception handlers for PL/pgSQL FWIW.

**Comment by mitchellwrosen** [comment]
Similar story for the `Transaction` newtype (private repo). It lacks a `MonadCatch` instance but does provide `transactionTry :: Transaction a -> Transaction (Either SqlError a)`, which takes care of creating a savepoint and rolling back.

**Comment by carymrobbins** [comment]
I think if a `SqlError` is raised from postgresql while in a transaction I think any other successive SQL calls will get another `SqlError` of the flavor `ERROR:  current transaction is aborted, commands ignored until end of transaction block`. Any attempt to `commit` should trigger a `rollback`.

```sql
𝔭𝔤 begin;
BEGIN

𝔭𝔤 create table t(a int);
CREATE TABLE

𝔭𝔤 insert into t select 1;
INSERT 0 1

𝔭𝔤 insert into t select 'a';
ERROR:  invalid input syntax for integer: "a"
LINE 1: insert into t select 'a';
                             ^

𝔭𝔤 insert into t select 'a';
ERROR:  current transaction is aborted, commands ignored until end of transaction block

𝔭𝔤 commit;
ROLLBACK
```

Of course, we should actually have a test case demonstrating this, but I think this might not be as concerning as originally thought.

**Comment by mitchellwrosen** [comment]
@carymrobbins precisely; that's why you'd want to rollback to the savepoint set before attempting the possibly-failing computation - so you _can_ continue the transaction

**Comment by jfischoff** [comment]
I wonder if should add a `txCatch` which can use savepoints with each backend? I don't know ... just spitballing. Otherwise each time `catch` is used users will have to add bespoke savepoint logic (assuming they care about recovering ... which I assume they would if they are catching the exception).

**Comment by jship** [comment]
A lot of times when we use `catch` in DB code, it's to figure out what the sql error is and then re-throw a friendlier error like a "not found".  I wonder if it's common to do a `catch` around DB code for recovery. 🤔

**Comment by jship** [comment]
Another option would be to punt on adding these `exceptions`-related instances for now, and allow people to catch exceptions with `Control.Exception` (in `IO`) via `unsafeWithRunInIOTxM`.  If we took that approach, it'd probably be good to make this unlift-style function available available not only for `TxM`, but also the underlying implementation monads like `PgQueryM`.

**Comment by jfischoff** [comment]
> A lot of times when we use catch in DB code, it's to figure out what the sql error is and then re-throw a friendlier error like a "not found"

Makes sense. Maybe we should just provide a `mapException` function? 

>  I wonder if it's common to do a catch around DB code for recovery.

It is not common because most transaction monads do not offer the ability to `catch`. However I added it to my monad and did find it useful.

**Comment by jship** [comment]
> Makes sense. Maybe we should just provide a mapException function?

Yeah I'll give that a go 👍

**Comment by carymrobbins** [comment]
Using `try` with `TxM` is going to be very tricky because it requires you to know something about the implementation. I could imagine that some sort of `try` could be provided in each of the implementation libs, e.g. `tryPgQueryM :: PgQueryM a -> PgQuery (Either e a)` but I don't think you can reliably create something to do `TxM a -> TxM (Either e a)` that has that savepoint behavior.

That being said, I think `mapException` or something similar is a good idea. I had also thought that something like `TxM a -> (e -> io Void) -> TxM a` might be useful too to easily re-throw, but `mapException` might be a bit more straightforward.

**Comment by jfischoff** [comment]
@carymrobbins yeah I figured getting the savepoint behavior would require separate implementations. Seems like kicking the can on that is best until we know we want it.

**Comment by jship** [comment]
Added `throwExceptionTx` and `mapExceptionTx` functions.  In my initial integration work of `postgresql-tx`, these functions seem to suit the integration's needs well.

File: src/Database/PostgreSQL/Tx.hs [hunk 1]
```diff
@@ -61,3 +66,30 @@ unsafeReaderIOTx
   :: (UnsafeTx (ReaderT r io) (ReaderT r t))
   => (r -> io a) -> ReaderT r t a
 unsafeReaderIOTx = unsafeIOTx . ReaderT
+
+throwExceptionTx :: (MonadIO io, UnsafeTx io t, Exception e) => e -> t a
+throwExceptionTx ex = unsafeIOTx $ liftIO $ throwIO ex
+
+mapExceptionTx
+  :: (UnsafeUnliftTx t, Exception e, Exception e')
+  => (e -> Maybe e')
```
Comment (carymrobbins) [inline src/Database/PostgreSQL/Tx.hs:75]: Is Maybe necessary? Seems like the user could just return the original exception if need be.

**Comment by jfischoff** [comment]
Is throwExceptionTx better than MonadThrow?

File: src/Database/PostgreSQL/Tx.hs (hunk 1 above)
Comment (jship) [reply src/Database/PostgreSQL/Tx.hs:75]: In the integration work I have this helper:

```haskell
fromPgViolation
  :: [PG.SqlError -> PG.ConstraintViolation -> Maybe Core.Error]
  -> PG.SqlError
  -> Maybe Core.Error
fromPgViolation toErrors sqlError = do
  violation <- PG.constraintViolation sqlError
  asum $ flip fmap toErrors \toError ->
    toError sqlError violation
````

which can be used like this:

```haskell
mapExceptionTx (fromPgViolation [some_table_pkey]) do
  -- blah blah
  where
  some_table_pkey :: PG.SqlError -> PG.ConstraintViolation -> Maybe Core.Error
  some_table_pkey sqlError violation = do
    guard (violation == violation'some_table_pkey)
    pure $ Core.buildError Core.Conflict $
      -- do stuff with either 'sqlError' or 'violation' to build meaningful error
```

So it's useful to go from one exception type to another, but converting from SqlError to a domain error isn't total, so went with `Maybe` in `mapExceptionTx`.

**Comment by jship** [comment]
> Is throwExceptionTx better than MonadThrow?

Nah I don't think it's bettter. Since I was whacking MonadCatch, I whacked MonadThrow too.  `throwExceptionTx` allows for throwing an exception straight in IO and without depending on `exceptions`, but no preference on my end for going through `throwExceptionTx` or through `exceptions`.

If you all prefer the `MonadThrow`, lemme know and I'll replace `throwExceptionTx` with the instance.

**Pull request merged**